platform-sdk/swirlds-base/src/main/java/com/swirlds/base/context/internal/ThreadLocalContext.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.base.context.internal;
+
+import com.swirlds.base.context.Context;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * The ThreadLocalContext is a {@link Context} implementation that defined as a singleton and is used to store
+ * information that is local to the current thread.
+ *
+ * @see GlobalContext
+ * @see Context
+ */
+public final class ThreadLocalContext implements Context {
+
+    private static final ThreadLocalContext INSTANCE = new ThreadLocalContext();
+
+    private final ThreadLocal<Map<String, String>> contextThreadLocal;
+
+    /**
+     * private constructor to prevent instantiation.
+     */
+    private ThreadLocalContext() {
+        this.contextThreadLocal = new ThreadLocal<>();
+    }
+
+    @Override
+    public AutoCloseable add(@NonNull String key, @NonNull String value) {
+        Objects.requireNonNull(key, "key must not be null");
+        Objects.requireNonNull(value, "value must not be null");
+
+        Map<String, String> contextMap = contextThreadLocal.get();
+        if (contextMap == null) {
+            contextMap = new HashMap<>();
+            contextThreadLocal.set(contextMap);
+        }
+        contextMap.put(key, value);
+        return () -> remove(key);
```

Comment:
What if somebody does something like this?

```
try (var x = context.add("foo", 123)) {
    System.out.println(getContextMap().get("foo")); // prints 123
    try (var y = context.add("foo", 321) {
        System.out.println(getContextMap().get("foo")); // prints 321
    }
    System.out.println(getContextMap().get("foo")); // prints null
}
```

We probably either need to create a stack of values for the context if such a thing is supported, or prevent nested contexts from overwriting the same value.

Reply:
I was already thinking about that. Currently my comment on that is like that:
I assume that the shown code is not a common use case. Since the context will contain info like transaction-id, etc. normally such call should not happen. But yes, I assume there are scenarios in that such code would happen. All other Context implementations (log4J2, SLF4J) have void methods but internally does the same. So they have the same problem. I assume that easy+fast code wins here since the context will be read in every log call. Maybe we will find a good stack based impl and future that is as fast or even faster as the current one. In that case we can change it in future. Next to that everything that is in the context is kind of optional metadata that will help us to interpret diagnostics (like logging). So the worst case is that such metadata will currently be missed in some special cases that might never happen.
I created https://github.com/hashgraph/hedera-services/issues/8457

Reply:
If we explicitly do not support this pattern, maybe we can just throw an exception if you try to add something that is already present.

Reply:
Why? I would say it is totally ok to just overwrite a value.